### PR TITLE
[Owners] System test for FetchMultiple API in dcpower

### PR DESCRIPTION
**Pull request by doshirohan**
### What does this Pull Request accomplish?

We recently added support for FetchMultiple API in dcpower. This pull request adds system test for the API.

### Why should this Pull Request be merged?

The system test will make sure that ViBoolean[] parameter support for FetchMultiple keeps working as expected since special handling was added for that in our mako scripts.

### What testing has been done?

- Executed the test locally.

### CODE REVIEW

File: source/tests/system/nidcpower_driver_api_tests.cpp [hunk 1]
```diff
@@ -447,6 +460,38 @@ TEST_F(NiDCPowerDriverApiTest, ConfigureOutputFunctionAndCurrentLevel_Configures
   EXPECT_EQ(expected_output_function_value, actual_output_function_value);
 }
 
+TEST_F(NiDCPowerDriverApiTest, FetchMultiple_ReturnsResultOfExpectedSize)
```
Comment (kruthi-NI) [inline source/tests/system/nidcpower_driver_api_tests.cpp]: Can given part of test name cover the steps of set attribute and initiate?

File: source/tests/system/nidcpower_driver_api_tests.cpp [hunk 2]
```diff
@@ -447,6 +460,38 @@ TEST_F(NiDCPowerDriverApiTest, ConfigureOutputFunctionAndCurrentLevel_Configures
...
+  const int expected_count = 20;
+  ::grpc::ClientContext context;
+  dcpower::FetchMultipleRequest request;
+  request.mutable_vi()->set_id(GetSessionId());
+  request.set_channel_name(channel_name);
+  request.set_timeout(timeout);
+  request.set_count(expected_count);
+  dcpower::FetchMultipleResponse response;
+
+  // Attribute 'NIDCPOWER_ATTRIBUTE_MEASURE_WHEN' must be set before calling FetchMultiple
```
Comment (kruthi-NI) [inline source/tests/system/nidcpower_driver_api_tests.cpp:480]: Extra lines should match given when then sections.

File: source/tests/system/nidcpower_driver_api_tests.cpp [hunk 3]
```diff
@@ -447,6 +460,38 @@ TEST_F(NiDCPowerDriverApiTest, ConfigureOutputFunctionAndCurrentLevel_Configures
...
 }
 
+TEST_F(NiDCPowerDriverApiTest, FetchMultiple_ReturnsResultOfExpectedSize)
+{
+  const char* channel_name = "0";
+  const int timeout = 5;
+  const int expected_count = 20;
+  ::grpc::ClientContext context;
+  dcpower::FetchMultipleRequest request;
+  request.mutable_vi()->set_id(GetSessionId());
```
Comment (kruthi-NI) [inline source/tests/system/nidcpower_driver_api_tests.cpp]: Fetch multiple API implementation can be part of helper method. Given section of code body can include setting of attribute and initiate

**Review by kruthi-NI — APPROVED** [review]

File: source/tests/system/nidcpower_driver_api_tests.cpp (hunk 1 above)
Comment (kruthi-NI) [reply source/tests/system/nidcpower_driver_api_tests.cpp]: Can test name be 
MeasurementIsInitiated_FetchMulitple_ReturnsResultOfExpectedSize

Comment (doshirohan) [reply source/tests/system/nidcpower_driver_api_tests.cpp]: Yep, just updated to `SetMeasureWhenAndInitiate_FetchMultiple_ReturnsResultOfExpectedSize`

File: source/tests/system/nidcpower_driver_api_tests.cpp (hunk 3 above)
Comment (doshirohan) [reply source/tests/system/nidcpower_driver_api_tests.cpp]: Great suggestion! Refactored test body to extract call to fetch multiple API into a function and now test is much more readable.

File: source/tests/system/nidcpower_driver_api_tests.cpp (hunk 2 above)
Comment (doshirohan) [reply source/tests/system/nidcpower_driver_api_tests.cpp:480]: Fixed.

File: source/tests/system/nidcpower_driver_api_tests.cpp [hunk 4]
```diff
@@ -70,6 +70,19 @@ class NiDCPowerDriverApiTest : public ::testing::Test {
     ASSERT_EQ(kdcpowerDriverApiSuccess, response.status());
   }
 
+  void initiate()
+  {
+      dcpower::InitiateRequest request;
```
Comment (reckenro) [inline source/tests/system/nidcpower_driver_api_tests.cpp]: [Minor] The spacing in the body of `initiate` and `fetch_multiple` is different than all the other methods. I think we're using two spaces per level of indent.

**Review by reckenro — APPROVED** [review]

File: source/tests/system/nidcpower_driver_api_tests.cpp [hunk 5]
```diff
@@ -447,6 +474,28 @@ TEST_F(NiDCPowerDriverApiTest, ConfigureOutputFunctionAndCurrentLevel_Configures
   EXPECT_EQ(expected_output_function_value, actual_output_function_value);
 }
 
+TEST_F(NiDCPowerDriverApiTest, SetMeasureWhenAndInitiate_FetchMultiple_ReturnsResultOfExpectedSize)
+{
+  const char* channel_name = "0";
+  const int expected_count = 20;
```
Comment (jonlugoatx) [inline source/tests/system/nidcpower_driver_api_tests.cpp]: How do you know the expected channel count is 20 when you're not in control of exactly which piece of DC power hardware will be present when this test is run?

The system tests are there just to make sure that you can transfer *some* data to/from the driver from the gRPC server. In order to test correctness of the data being returned or any corner cases, you'll need to write an integration test using the mock library.

**Review by jonlugoatx — CHANGES_REQUESTED** [review]

File: source/tests/system/nidcpower_driver_api_tests.cpp (hunk 5 above)
Comment (doshirohan) [reply source/tests/system/nidcpower_driver_api_tests.cpp]: In this particular case, I believe unit test would be sufficient too. Only "outside" interaction [FetchMulitple](https://github.com/doshirohan/grpc-device/blob/features/ni-dcpower/generated/nidcpower/nidcpower_service.cpp#L2171) has is with `SessionRepository` and our [ni-fake unit test ](https://github.com/doshirohan/grpc-device/blob/features/ni-dcpower/source/tests/unit/ni_fake_service_tests.cpp#L45)seems to be using SessionRespository instance as well. Thoughts?

Comment (doshirohan) [reply source/tests/system/nidcpower_driver_api_tests.cpp]: Actually now that I think about it, the unit test added against ni-fake metadata does take care of validating that API returns correct values (conversion from ViBoolean[] to bool[] worked as expected). That's what essentially I was trying to do with expected_count check here. I will update this system test just to check if API returned success and let existing unit test worry about validity of data.

File: source/tests/system/nidcpower_driver_api_tests.cpp (hunk 4 above)
Comment (doshirohan) [reply source/tests/system/nidcpower_driver_api_tests.cpp]: Ah, I need to update my VS settings. Fixed the spacing.

File: source/tests/system/nidcpower_driver_api_tests.cpp (hunk 5 above)
Comment (doshirohan) [reply source/tests/system/nidcpower_driver_api_tests.cpp]: I have removed assumptions regarding size and we will rely on unit test [NiFakeService_BoolArrayOutputFunction_CallsBoolArrayOutputFunction](https://github.com/doshirohan/grpc-device/blob/features/ni-dcpower/source/tests/unit/ni_fake_service_tests.cpp#L501) on nifake to verify integrity of data.

**Review by jonlugoatx — APPROVED** [review]
